### ISSUE TITLE
Confirm Signatures and Finalizing UTXO positions

### ISSUE THREAD

**Issue by colin-axner** [FourthState/plasma-mvp-sidechain#10]
In order to verify that a utxo being spent is valid we need check either:
1. UTXO positions match up with the current state of our utxo database
2. ConfirmSig is valid
or we need to check:
1. UTXO referenced existed in the block it is trying to be spent from (would need merkle proof)
2. ConfirmSig is valid

In order to use the first method we would need to update our utxo database after a block is submitted. This is currently not possible using CheckTx and DeliverTx since we are missing the information for the transaction index. Currently we use a finalize message to be passed around to update the database. However, there are two issues with this:
1. FinalizeMsg is included into blocks and would require a fee
2. FinalizeMsg would probably need to supply a merkle proof to be verified

If we wanted to verify the existence of a UTXO using a merkle proof, we would need access to our cms (commitMultiStore) in base app. This is only accessible in baseapp. We would also need to encode merkle proofs into transactions (how large is our transaction size becoming?). Perhaps we could add a KVStore of merkle roots to our context which we could add when endBlock() is called? If this is possible then we could add a merkle proof to our spendMsg and the ante handler should have enough information to verify its existence.  Perhaps we adjust our utxo database to hold all spentMsg's to protect against double spends?
 
In summary:
- Add merkle proof to spendMsg
- Add merkle proof mapper to ctx
- Ante handler check existence of utxo inputs
- Utxo db holds spent tx's to protect against double spend

Thoughts?

**Comment by AdityaSripal** [comment]
To verify a UTXO being spent is valid:
If it is a deposit:
Form the corresponding transaction that created the deposit and check if hash matches with rootchain. Also check that this transaction is not in the Spent Deposit store.

If it is not a deposit:
Check that the UTXO exists in the (Position => UTXO) UTXO store.

For both cases:
Verify that the signature matches address in Owner field and denom matches claimed amount

**Comment by colin-axner** [comment]
#25
